### DISCUSSION TITLE
Return the output of `onValid` and `onInvalid` when `handleSubmit` returns

### DISCUSSION THREAD

**Discussion by petercpwong** [react-hook-form/react-hook-form#8590] [Ideas]
Currently there is no way to inspect the output of the `onValid` or `onInvalid` functions after it finishes running inside `handleSubmit`.

Saving the output of `onValid` and `onInvalid` and returning it after `handleSubmit` returns would be immensely helpful.

https://github.com/react-hook-form/react-hook-form/blob/efd9195b179f14f6f256478c91d81b6bf7b0da3c/src/logic/createFormControl.ts#L1013-L1067

One use case may be for another downstream hook to determine the next steps based on the output of `onValid` (eg. a Modal window hook that returns server data from `onValid` after it closes).  Anything downstream running the function returned by `handleSubmit` is unable to inspect the output of `onValid` since whatever it returns gets swallowed. 

Example use case:

```javascript
function useModal({ onSubmit }) {
  const submitModal = useCallback(() => {
    // No way to inspect what was returned by handleSubmit
    const result = await onSubmit();
    
    if (result.foo) {
      // Do something interesting
    }
  }, [onSubmit]);

  return {
    submitModal
  }
}


/* Inside some component */

const { handleSubmit } = useForm();

const submit = handleSubmit(async () => {
  const result = await someHttpRequest();
  // This gets swallowed
  return result;
});

useModal({ onSubmit: submit });
```

**Comment by bluebill1049** [comment]
Did you see my proposal on the oringal feature reques? Let me know what do you think?

**Reply by petercpwong** [reply]
> Did you see my proposal on the oringal feature reques? Let me know what do you think? 

Are you referring to this other thread:

https://github.com/react-hook-form/react-hook-form/issues/7661#issuecomment-1112766670

**Reply by bluebill1049** [reply]
yes, does that help your problem? Sorry i haven't got time to look closer to your question above, just quickly scanned will take a closer look after work or in the weekend.

**Reply by petercpwong** [reply]
Yes the proposed changes in the other thread look good. Much better than what I'd originally proposed.

**Reply by petercpwong** [reply]
@bluebill1049 Can we revisit this PR again? The other feature request veered off into a different use case from the one this PR attempts to solve. This PR allows the user to (optionally) "peek" into the return value of `handleSubmit` while the other [recent feature](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.43.0-next.0) allows the user to emit global errors.

**Reply by bluebill1049** [reply]
hey, @petercpwong I am working on the new Form component, what if we include values in the `onScucess` callback?

```tsx
<Form
  onSuccess={({ data }) => {
    console.log(data)
  }}
/>
``

**Reply by petercpwong** [reply]
My use case requires being able to get the return value from the function created by `handleSubmit`.

Also looks like other people have a similar use case:
https://github.com/react-hook-form/react-hook-form/pull/6817#issuecomment-1413227332

It's a very small non-breaking change that would help me (and possibly many others) tremendously.

**Reply by bluebill1049** [reply]
I see, sure let me revisit this once Form component is stable for a beta release.

**Comment by cb-eli** [comment]
It will be helpful on my side as well. Currently, I'm using refs as a dirty workaround.
